src/ontology/OntologyAnnotatedTypeFactory.java

```diff
@@ -54,74 +46,109 @@ public TreeAnnotator createTreeAnnotator() {
-                Object... args) {
-            super.finish(qualHierarchy, fullMap, polyQualifiers, tops, bottoms, args);
-
-            // substitue ONTOLOGY with ONTOLOGY_TOP in fullMap
-            assert fullMap.containsKey(OntologyUtils.ONTOLOGY);
-            Set<AnnotationMirror> ontologyTopSupers = fullMap.get(OntologyUtils.ONTOLOGY);
-            fullMap.remove(OntologyUtils.ONTOLOGY);
-            fullMap.put(OntologyUtils.ONTOLOGY_TOP, ontologyTopSupers);
-
-            // update tops
-            tops.remove(OntologyUtils.ONTOLOGY);
-            tops.add(OntologyUtils.ONTOLOGY_TOP);
+        public @Nullable AnnotationMirror leastUpperBound(
+                AnnotationMirror a1, AnnotationMirror a2) {
+            if (getQualifierKind(a1) != ONTOLOGY_KIND || getQualifierKind(a2) != ONTOLOGY_KIND) {
+                throw new BugInCF("unexpected annotation mirrors: %s, %s", a1, a2);
+            }
+
+            EnumSet<OntologyValue> a1Set = OntologyUtils.getOntologyValuesSet(a1);
+            EnumSet<OntologyValue> a2Set = OntologyUtils.getOntologyValuesSet(a2);
+
+            return OntologyUtils.createOntologyAnnotationByValues(
+                    processingEnv,
+                    OntologyUtils.lubOfOntologyValues(a1Set, a2Set)
+                            .toArray(new OntologyValue[] {}));
         }
 
         @Override
-        public boolean isSubtype(AnnotationMirror rhs, AnnotationMirror lhs) {
-            if (AnnotationUtils.areSameByName(rhs, OntologyUtils.ONTOLOGY)
-                    && AnnotationUtils.areSameByName(lhs, OntologyUtils.ONTOLOGY)) {
-                OntologyValue[] rhsValue = OntologyUtils.getOntologyValues(rhs);
-                OntologyValue[] lhsValue = OntologyUtils.getOntologyValues(lhs);
-                EnumSet<OntologyValue> rSet = EnumSet.noneOf(OntologyValue.class);
-                rSet.addAll(Arrays.asList(rhsValue));
-                EnumSet<OntologyValue> lSet = EnumSet.noneOf(OntologyValue.class);
-                lSet.addAll(Arrays.asList(lhsValue));
-
-                if (rSet.containsAll(lSet)
-                        || rSet.contains(OntologyValue.BOTTOM)
-                        || lSet.contains(OntologyValue.TOP)) {
-                    return true;
-                } else {
-                    return false;
-                }
-            } else {
-                return super.isSubtype(rhs, lhs);
+        public @Nullable AnnotationMirror greatestLowerBound(
+                AnnotationMirror a1, AnnotationMirror a2) {
+            if (getQualifierKind(a1) != ONTOLOGY_KIND || getQualifierKind(a2) != ONTOLOGY_KIND) {
+                throw new BugInCF("unexpected annotation mirrors: %s, %s", a1, a2);
             }
+
+            Set<OntologyValue> a1Set = OntologyUtils.getOntologyValuesSet(a1);
+            Set<OntologyValue> a2Set = OntologyUtils.getOntologyValuesSet(a2);
+
+            if (a1Set.contains(OntologyValue.POLY) && a2Set.contains(OntologyValue.POLY)) {
```

Comment:
Should this logic go into a method similar to `OntologyUtils.lubOfOntologyValues`?

Reply:
Sure, I think it makes the code look more symmetric. But so far this is the only place that needs GLB.